vnext/Microsoft.ReactNative.Cxx/PropertyBag.h

```diff
@@ -0,0 +1,249 @@
+    return m_namespace ? m_namespace.NamespaceName() : hstring{};
+  }
+
+  IReactPropertyNamespace const &Get() const noexcept {
+    return m_namespace;
+  }
+
+  explicit operator bool() const noexcept {
+    return static_cast<bool>(m_namespace);
+  }
+
+ private:
+  IReactPropertyNamespace m_namespace;
+};
+
+//! Encapsulates the IReactPropertyName and the property type
+template <class T>
+struct PropertyName {
+  using PropertyType = T;
+
+  PropertyName(std::nullptr_t = nullptr) noexcept {}
+
+  explicit PropertyName(IReactPropertyName const &name) noexcept : m_name{name} {}
+
+  explicit PropertyName(hstring const &localName) noexcept : m_name{ReactPropertyBag::GetName(nullptr, localName)} {}
+
+  PropertyName(PropertyNamespace const &ns, hstring const &localName) noexcept
+      : m_name{ReactPropertyBag::GetName(ns.Get(), localName)} {}
+
+  PropertyName(hstring const &ns, hstring const &localName) noexcept
+      : m_name{ReactPropertyBag::GetName(ReactPropertyBag::GetNamespace(ns), localName)} {}
+
+  hstring NamespaceName() const noexcept {
+    return m_name ? m_name.Namespace().NamespaceName() : hstring{};
+  }
+
+  hstring LocalName() const noexcept {
+    return m_name ? m_name.LocalName() : hstring{};
+  }
+
+  IReactPropertyName const &Get() const noexcept {
+    return m_name;
+  }
+
+  explicit operator bool() const noexcept {
+    return static_cast<bool>(m_name);
+  }
+
+ private:
+  IReactPropertyName m_name;
+};
+
+//! PropertyBag is a wrapper for IReactPropertyBag that stores strongly-typed properties in a thread-safe way.
+//! Types inherited from IInspectable are stored directly.
+//! Values of other types are boxed with help of winrt::box_value.
+struct PropertyBag {
+  //! Property return type is either T or std::optional<T>.
+  //! T is returned for types inherited from IInspectable.
+  //! The std::optional<T> is returned for all other types.
+  template <class T>
+  using ResultType = std::conditional_t<std::is_base_of_v<Windows::Foundation::IInspectable, T>, T, std::optional<T>>;
+
+  //! Create a new empty instance of PropertyBag.
+  PropertyBag(std::nullptr_t = nullptr) noexcept {}
+
+  //! Creates a new instance of PropertyBag with the provided content.
+  explicit PropertyBag(IReactPropertyBag const &content) noexcept : m_content{content} {}
+
+  //! True if content is not null.
+  explicit operator bool() const noexcept {
+    return static_cast<bool>(m_content);
+  }
+
+  //! Get content from the PropertyBag.
+  IReactPropertyBag const &Get() const noexcept {
+    return m_content;
+  }
+
+  template <class T>
+  static Windows::Foundation::IInspectable ToObject(T const &value) noexcept {
+    return winrt::box_value(value);
+  }
+
+  template <class T>
+  static Windows::Foundation::IInspectable ToObject(std::optional<T> const &value) noexcept {
+    return value ? ToObject(*value) : nullptr;
+  }
+
+  template <class T>
+  static auto FromObject(Windows::Foundation::IInspectable const &obj) noexcept {
```

Comment:
Could you add a comment to these or designate them as private?

Reply:
Good point! I meant these methods to be private.